### ISSUE TITLE
[labs/analyzer] extract generic analyses from lit-element logic

### ISSUE THREAD

**Issue by 43081j** [lit/lit#3521]
### Should this be an RFC?

- [X] This is not a substantial change

### Which package is this a feature request for?

Other/unknown (please mention in description)

### Description

Currently the analyzer can analyze roughly the following from an element:

* Lit decorators (properties, custom element definition)
* JSDoc tags (events, slots, css properties, etc)
* `customElements.define` calls

Ultimately, this stuff only runs right now if we get a truthy `isLitElementSubclass` result (i.e. something which inherits `LitElement`).

That basically means the analyzer **can only currently analyzer lit elements**.

Most of the logic I mentioned above, though, is _not_ specific to lit elements. I wonder if we could actually split it up into two things:

* Anything which inherits `HTMLElement` (including lit elements)
  * JSDoc tags
  * `customElements.define` calls
* Anything which inherits `LitElement`
  * Lit decorators (properties, customElement)

This'd be fairly easy by doing a few things:

* Create an `isCustomElementSubclass` type guard
* Create a `CustomElementDeclaration` model which actually holds everything the current `LitElementDeclaration` model has _except_ reactive properties
* Update `LitElementDeclaration` to inherit the new `CustomElementDeclaration`
* Split the tag name logic up so the custom element declaration looks for define calls, and the lit declaration does that _in addition to_ checking for `@customElement` decorators

Then shuffle the tests around a bit. Somehow structure it in a way that means we can, for example, run the JSDoc tests against a vanilla element _and_ a lit element with little code duplication.

clearly i could have a stab at this myself but i think i'd get stuck in a rabbit hole with how to nicely rework the tests.

### Alternatives and Workarounds

N/A

**Comment by justinfagnani** [comment]
@43081j I think we could consider this, but to be clear about this intention of the analyze: it's mostly a non-goal to be a generic custom element analyzer, especially one with any kind of plug in system. We want something we can iterate fast on and keep perfectly in sync and changes to Lit syntax or semantics by being in the monorepo and sharing tests.

Generic custom element analysis _might_ be best served by something like Open WC's CEM Analyzer.

That said, I do acknowledge that the model of elements that we build up has only a few Lit specifics parts to it. It seems like it wouldn't be too disruptive to the primary goal of supporting Lit to layer the models so that there's a generic layer and a Lit specialization. From that we might support vanilla custom elements with jsdoc syntax and other WC libraries could consume the analyzer as a library to build their own analyzer on.

I'll let @kevinpschaaf weight in from here, since he's the on primarily working on this now.

**Comment by 43081j** [comment]
Yeah I do understand that and if I was in your position I'd be tempted to keep it tightly scoped to lit too. 

My thought was that lit inherits custom element behaviours anyway so those two layers being supported might be ok. As long as you don't end up expanding that to other non lit layers. 

It's not an easy one to decide though still. On one hand you have some quality implementation of something I can see would be useful to people. On the other hand you really don't want to end up supporting some generic analyzer a bunch of loosely related tools rely on. 

I guess if you think of a language service, manifest generator, linter, etc. Would you want people to have to install two of each of those tools to support their lit sources and non-lit dependencies. Or would you cover both ends yourself

**Comment by kevinpschaaf** [comment]
Yeah, James and I chatted a bit earlier.

While we don't plan on supporting a plugin system to support non-Lit framework-specific idioms, if a class is `@customElement()`-decorated or `customElements.define()`ed, we can trivially include the generic analysis of the class as a `CustomElementDeclaration`.

That said, it _is_ a goal for our analyzer to fall back to using CEM for imported packages. So you should still get roughly the same internal analyzer model for imported non-Lit components as for Lit components being analyzed from source -- and I think this is the sweet spot we're shooting for.

In other words, we only expect the analyzer to be used to analyze raw source code for actual (Lit) code under development, not dependencies -- since those dependencies will (in the long run) all come with their own CEM, making source-code analysis of those unnecessary. As long as a developer isn't mixing-matching the library they're using to build new components, they can mix-match components built with different libraries all day long, and there shouldn't be a need to install more than one LSP, linter, etc.

**Comment by 43081j** [comment]
That makes sense I think  

Presumably this would mean the analyzer at some point needs to understand manifests? So once it passes the boundary of the local project and enters an external module, it looks for a manifest and understands it that way.

**Comment by kevinpschaaf** [comment]
Yup, that's exactly right. In my [prototype](https://github.com/lit/lit/compare/analyzer-pull/#diff-39757f0b8cef9fd09964c798cc454ad0fc70fc019a56b3db1b1d87adda4425fdR166) last summer (that I'm slowly but surely trying to productionze) the logic basically went 
```ts
if (isSourceFileInProgram(sourceFile)) {
  /* analyze from source */
} else {
  importPackageFromManifest(packageRoot, packageJson);
}
```

One modification to this that @justinfagnani pointed out is that _if_ there is a `.d.ts` file for the imported module, there may be some benefit to merging info from the `.d.ts` file and the `custom-elements.json` manifest together; this would potentially allow getting at the TS-based types for e.g. type checking bindings to class fields, along with all the CE-specific stuff coming from the manifest. That needs a little more design/investigation.
